VnodeCacheDesign.md

```diff
@@ -0,0 +1,55 @@
+# Vnode Cache Design
+
+## Goals
+
+- Reduce the time required to determine if a file is inside of a virtualization root
+- Reduce the time required to find a file's virtualization root
+- **TBD:** Eliminate the use of file flags to determine if a file is inside of a root
+
+## Design 
+
+### Cached Data
+
+PrjFSKext will cache the following information for each vnode:
+
+- The `vnode_t` itself
+- The vnode's `uint32_t` vid (generation number)
+- `VirtualizationRootHandle` for the vnode (or `RootHandle_None` if the vnode is not in a root).  **NOTE** This assumes that `VirtualizationRootHandle`  are not recycled as VFS4G instances are unmounted\remounted.
```

Comment:
I think that is a safe assumption. A "handle" here is really just an index into the array. We should make sure that we only ever expand that array, and never move entries around. That is its current behavior, but some comments / unit tests would help reinforce that.

Reply:
We'll probably want to recycle indices if a root is deleted/destroyed. However, in that case we should be able to ensure the cache does not refer to that index anymore.

Reply:
> Presumably we'll use RootHandle_Indeterminate as the "unknown" (invalidated cache) state.

When portions of the cache are invalidated we won't need to update the `VirtualizationRootHandle` member of any of the entries in the cache.  We'll know they are invalidated because their `vrgid` won't match the current `vrgid` for their `VirtualizationRootHandle`.